### ISSUE TITLE
Invalid free in hs_alloc_scratch()

### ISSUE THREAD

**Issue by hlippmann** [intel/hyperscan#174]
In an out-of-memory situation the function _hs_alloc_scratch()_ calls _hs_scratch_free()_ with an invalid parameter. This may lead to an invalid _free()_-call making the calling application crash.

Here is the buggy code _hs_alloc_scratch()_ in scratch.c:283:

```c
    hs_scratch_t *proto;
    hs_scratch_t *proto_tmp = hs_scratch_alloc(sizeof(struct hs_scratch) + 256);
    hs_error_t proto_ret = hs_check_alloc(proto_tmp);
    if (proto_ret != HS_SUCCESS) {
        hs_scratch_free(proto_tmp);
        hs_scratch_free(*scratch);       <------------------- BUG!
        *scratch = NULL;
        return proto_ret;
    }
```

Verified with version 5.1.1, but older versions seem to be affected as well.

The following fix works for me:
```patch
diff --git a/src/scratch.c b/src/scratch.c
index c23b5b3..ae8fa4c 100644
--- a/src/scratch.c
+++ b/src/scratch.c
@@ -280,7 +280,7 @@ hs_error_t HS_CDECL hs_alloc_scratch(const hs_database_t *db,
     hs_error_t proto_ret = hs_check_alloc(proto_tmp);
     if (proto_ret != HS_SUCCESS) {
         hs_scratch_free(proto_tmp);
-        hs_scratch_free(*scratch);
+        hs_scratch_free((*scratch)->scratch_alloc);
         *scratch = NULL;
         return proto_ret;
     }
```

**Comment by xiangwang1** [comment]
What do you mean by `invalid parameter`? 
I think it should be fine to free a NULL pointer. Any particular reason why freeing `(*scratch)->scratch_alloc` succeeds while freeing (*scratch) fails?

**Comment by hlippmann** [comment]
You are right, freeing a NULL ptr is fine, but this is not what's gonna happen here. Here _scratch_ has a proper value, but, as far as I can see, this is not the original memory address that was returned by allocator beforehand. This one is kept in _(*scratch)->scratch_alloc_ instead.

If you see a few lines below (line 372) in _scratch.c_ you see that there exactly the same _hs_scratch_free()_ call as in my patch is used.

**Comment by xiangwang1** [comment]
Thanks for the explanation! We'll fix it in the code.

**Comment by Nor7th** [comment]
Fixed in 5.2.0.
